In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats.stats import pearsonr
import scipy as sp
from scipy import stats
from scipy.stats import skew,norm
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import PolynomialFeatures
plt.style.use('seaborn')
from sklearn import metrics

In [40]:
df_corr_matrix = df.corr() 
df_corr_matrix  #computed a correlation matrix for data set to see if it suggested that any features were collinear

carat       cut     color   clarity     depth     table  \
carat      1.000000 -0.134967 -0.291437 -0.352841  0.028224  0.181618   
cut       -0.134967  1.000000  0.020519  0.189175 -0.218055 -0.433405   
color     -0.291437  0.020519  1.000000 -0.025631 -0.047279 -0.026465   
clarity   -0.352841  0.189175 -0.025631  1.000000 -0.067384 -0.160327   
depth      0.028224 -0.218055 -0.047279 -0.067384  1.000000 -0.295779   
table      0.181618 -0.433405 -0.026465 -0.160327 -0.295779  1.000000   
price      0.921591 -0.053491 -0.172511 -0.146800 -0.010647  0.127134   
Length_mm  0.975094 -0.125565 -0.270287 -0.371999 -0.025289  0.195344   
Width_mm   0.951722 -0.121462 -0.263584 -0.358420 -0.029341  0.183760   
Depth_mm   0.953387 -0.149323 -0.268227 -0.366952  0.094924  0.150929   
Ratio      0.118238 -0.024523 -0.031592 -0.126955  0.049873  0.121812   

              price  Length_mm  Width_mm  Depth_mm     Ratio  
carat      0.921591   0.975094  0.951722  0.953387  0.118238  
cut       -0.053491  -0.125565 -0.121462 -0.149323 -0.024523  
color     -0.172511  -0.270287 -0.263584 -0.268227 -0.031592  
clarity   -0.146800  -0.371999 -0.358420 -0.366952 -0.126955  
depth     -0.010647  -0.025289 -0.029341  0.094924  0.049873  
table      0.127134   0.195344  0.183760  0.150929  0.121812  
price      1.000000   0.884435  0.865421  0.861249  0.074599  
Length_mm  0.884435   1.000000  0.974701  0.970772  0.141480  
Width_mm   0.865421   0.974701  1.000000  0.952006  0.001633  
Depth_mm   0.861249   0.970772  0.952006  1.000000  0.112408  
Ratio      0.074599   0.141480  0.001633  0.112408  1.000000

In [185]:
mlr_model = ols(formula = "price ~ carat + cut + color + clarity", data = df).fit()  #Without scaling the numbers
mlr_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.904
Model:                            OLS   Adj. R-squared:                  0.904
Method:                 Least Squares   F-statistic:                 1.272e+05
Date:                Fri, 08 May 2020   Prob (F-statistic):               0.00
Time:                        02:45:25   Log-Likelihood:            -4.6053e+05
No. Observations:               53940   AIC:                         9.211e+05
Df Residuals:                   53935   BIC:                         9.211e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -7218.5545     33.296   -216.799      0.000   -7283.815   -7153.294
carat       8783.7718     12.692    692.087      0.000    8758.896    8808.648
cut          155.6999      4.863     32.014      0.000     146.167     165.232
color        319.6727      3.302     96.808      0.000     313.200     326.145
clarity      524.8435      3.527    148.799      0.000     517.930     531.757
==============================================================================
Omnibus:                    10789.064   Durbin-Watson:                   0.936
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           251931.502
Skew:                           0.366   Prob(JB):                         0.00
Kurtosis:                      13.562   Cond. No.                         48.2
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [48]:
#Scaling the Model so the features are represented in standard deviations as opposed to it's normal numbers
scaler = StandardScaler()
scaler.fit(df)
df_scaled = scaler.transform(df)
df_scaled = pd.DataFrame(df_scaled)
df_scaled.columns = list(df.columns)

In [50]:
mlr_model = ols(formula = "price ~ carat + cut + color + clarity", data = df_scaled).fit()
mlr_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.904
Model:                            OLS   Adj. R-squared:                  0.904
Method:                 Least Squares   F-statistic:                 1.272e+05
Date:                Thu, 07 May 2020   Prob (F-statistic):               0.00
Time:                        16:57:57   Log-Likelihood:                -13288.
No. Observations:               53940   AIC:                         2.659e+04
Df Residuals:                   53935   BIC:                         2.663e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -2.359e-16      0.001  -1.77e-13      1.000      -0.003       0.003
carat          1.0437      0.002    692.087      0.000       1.041       1.047
cut            0.0436      0.001     32.014      0.000       0.041       0.046
color          0.1363      0.001     96.808      0.000       0.134       0.139
clarity        0.2167      0.001    148.799      0.000       0.214       0.220
==============================================================================
Omnibus:                    10789.064   Durbin-Watson:                   0.936
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           251931.502
Skew:                           0.366   Prob(JB):                         0.00
Kurtosis:                      13.562   Cond. No.                         1.70
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [57]:
target = df_scaled.price
features = df_scaled.drop(["price", "Ratio"], axis = 1) #dropped Ratio in addition because I realized it wasn't as useful

In [58]:
X_train, X_test, y_train, y_test = train_test_split(features, target, train_size = 0.8, test_size = 0.2, random_state = 15)

In [54]:
X_train.columns

Index(['carat', 'cut', 'color', 'clarity', 'depth', 'table', 'Length_mm',
       'Width_mm', 'Depth_mm'],
      dtype='object')

In [62]:
## training the model
from sklearn.linear_model import Lasso

lasso = Lasso(alpha = 0.01, normalize = False)

lasso = lasso.fit(X_train, (y_train))

y_train_pred = lasso.predict(X_train)

train_mae = metrics.mean_absolute_error(y_train, (y_train_pred))  #getting the mean absolute error
train_mse = metrics.mean_squared_error(y_train, (y_train_pred))   #getting the mean squared error
train_rmse = np.sqrt(metrics.mean_squared_error(y_train, (y_train_pred)))  #getting the root mean squared error

In [63]:
print(train_mae)
print(train_mse)
print(train_rmse)

0.21042431537527861
0.09559166947418488
0.309179024958332


In [66]:
print ("R^2 Score:", lasso.score(X_train, (y_train)))  #getting the R^2 value

R^2 Score: 0.9042486341262427


In [65]:
y_pred = lasso.predict(X_test)

In [75]:
test_mae = metrics.mean_absolute_error(y_test, y_pred)
test_mse = (metrics.mean_squared_error(y_test, y_pred))
test_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_pred))

In [76]:
print(test_mae)
print(test_mse)
print(test_rmse)

0.21317011521713228
0.09967541209980615
0.3157141303454854


In [69]:
## training the model
from sklearn.linear_model import LinearRegression

lm = LinearRegression()

lm = lm.fit(X_train, (y_train))


y_train_pred = lm.predict(X_train)

train_mae = metrics.mean_absolute_error(y_train, (y_train_pred))  #getting the mean absolute error
train_mse = metrics.mean_squared_error(y_train, (y_train_pred))   #getting the mean squared error
train_rmse = np.sqrt(metrics.mean_squared_error(y_train, (y_train_pred)))  #getting the roort mean squared error

In [70]:
print(train_mae)
print(train_mse)
print(train_rmse)

0.20123486730377438
0.09209373418844573
0.3034694946587642


In [116]:
print ("R^2 Score:", lm.score(X_train, (y_train)))  #getting the R^2 value

R^2 Score: 0.9077524130976726


In [117]:
y_pred = lm.predict(X_test)

In [118]:
test_mae = metrics.mean_absolute_error(y_test, y_pred)
test_mse = (metrics.mean_squared_error(y_test, y_pred))
test_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_pred))

In [119]:
print(test_mae)
print(test_mse)
print(test_rmse)

0.20438958057244488
0.09709165628967366
0.3115953406096979


In [187]:
sd = test_rmse / y_test.std()  #shows how many standard deviations the average error is away from the true value
sd

0.31054980576001195

In [77]:
lm.coef_ #wanted to see the different coefficients from my features 

array([ 1.27718005,  0.03467103,  0.1375983 ,  0.20645464, -0.02883154,
       -0.01433144, -0.29421679,  0.05709798, -0.00200008])

In [114]:
df2 = df.drop(["price", "Ratio"], axis = 1)  #creating a new Data Frame without my target or the column I created
df2.head()

carat  cut  color  clarity  depth  table  Length_mm  Width_mm  Depth_mm
1   0.23    5      6        2   61.5   55.0       3.95      3.98      2.43
2   0.21    4      6        3   59.8   61.0       3.89      3.84      2.31
3   0.23    2      6        5   56.9   65.0       4.05      4.07      2.31
4   0.29    4      2        4   62.4   58.0       4.20      4.23      2.63
5   0.31    2      1        2   63.3   58.0       4.34      4.35      2.75

In [198]:
#creating new polynomial features

from sklearn.preprocessing import PolynomialFeatures

polynomial_features = PolynomialFeatures(degree = 2)
x_poly = polynomial_features.fit_transform(df2)

polynomial_features = PolynomialFeatures(degree = 2)
x_poly = polynomial_features.fit(df2)
x_poly_df = pd.DataFrame(x_poly.transform(df2),columns = x_poly.get_feature_names(df2.columns))

In [202]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(x_poly_df, target, train_size = 0.8, test_size = 0.2, random_state = 15)

In [203]:
from sklearn.linear_model import LinearRegression

lm_poly = LinearRegression()

lm_poly = lm.fit(X_train2, (y_train2))


y_train_p = lm_poly.predict(X_train2)
y_test_p = lm_poly.predict(X_test2)

train_mae = metrics.mean_absolute_error(y_train2, (y_train_p))  #getting the mean absolute error
train_mse = metrics.mean_squared_error(y_train2, (y_train_p))   #getting the mean squared error
train_rmse = np.sqrt(metrics.mean_squared_error(y_train2, (y_train_p)))  #getting the roort mean squared error

In [204]:
train_mae = metrics.mean_absolute_error(y_train2, (y_train_p))
train_mse = metrics.mean_squared_error(y_train2, (y_train_p))
train_rmse = np.sqrt(metrics.mean_squared_error(y_train2, (y_train_p)))

In [205]:
print(train_mae)
print(train_mse)
print(train_rmse)

0.11382011498108593
0.0356876433444278
0.1889117342687526


In [206]:
print ("R^2 Score:", lm_poly.score(X_train2, (y_train2)))

R^2 Score: 0.9642527365214887


In [207]:
test_mae = metrics.mean_absolute_error(y_test2, y_test_p)
test_mse = (metrics.mean_squared_error(y_test2, y_test_p))
test_rmse = np.sqrt(metrics.mean_squared_error(y_test2, y_test_p))

In [208]:
print(test_mae)
print(test_mse)
print(test_rmse)

0.12948847379656558
2.0430209538550668
1.4293428398586068


In [209]:
x_poly_df.columns  #wanted to see all the new features

Index(['1', 'carat', 'cut', 'color', 'clarity', 'depth', 'table', 'Length_mm',
       'Width_mm', 'Depth_mm', 'carat^2', 'carat cut', 'carat color',
       'carat clarity', 'carat depth', 'carat table', 'carat Length_mm',
       'carat Width_mm', 'carat Depth_mm', 'cut^2', 'cut color', 'cut clarity',
       'cut depth', 'cut table', 'cut Length_mm', 'cut Width_mm',
       'cut Depth_mm', 'color^2', 'color clarity', 'color depth',
       'color table', 'color Length_mm', 'color Width_mm', 'color Depth_mm',
       'clarity^2', 'clarity depth', 'clarity table', 'clarity Length_mm',
       'clarity Width_mm', 'clarity Depth_mm', 'depth^2', 'depth table',
       'depth Length_mm', 'depth Width_mm', 'depth Depth_mm', 'table^2',
       'table Length_mm', 'table Width_mm', 'table Depth_mm', 'Length_mm^2',
       'Length_mm Width_mm', 'Length_mm Depth_mm', 'Width_mm^2',
       'Width_mm Depth_mm', 'Depth_mm^2'],
      dtype='object')

In [182]:
len(x_poly_df.columns)  #wanted to see the length of all the new features

55

My approach to this project was to first find a data set of diamonds that I could work with that had enough columns(features) for me to engineer with so that I could make the best possible prediction model.  I had some prior background knowledge on diamonds and jewelry in general, however I wanted to put that to the test to see if my predictions of certain features would directly influence my target variable which was price.  I was also keeping in mind the entire time the different types of graphs I could use to help visualize my various correlations.  

The final model I decided to use to best predict the price of diamonds is my OLS Model.  My R^2 value was very similar for all four of my models however my OLS Model had the least amount of features which is ideal for a better fitting prediction model.  One feature that had a significant impact on our dependent variable of price was color of the diamond.  The coefficient for color was 0.1363 standard deviations which is a strong number.  For every grade of color that went up the price would increase by approximately 319.67 dollars.  A second feature that had a significant impact on price was clarity of the diamond.  The coefficient for clarity was 0.2167 standard deviations which is also a strong number.  For every grade of clarity that went up the price would increase by roughly 524.84 dollars.  The final and most significant feature that effected price was carat weight of the diamond.  The coefficient for carat was 1.0437 standard deviations which is the highest of the other two and all other features from this data set.  For every increase in carat weight, the price goes up by about 8,783.77 dollars.  All of these features had a P value under 0.05 which means that they are all statistically significant.  From this, we can say that we are 95 percent confident that our data is valuable the model because it effects price.